File: skills/optimization/reference/src/result-types.ts
/**
 * Result types and configuration for optimization algorithms.
 *
 * Defines the OptimizeResult returned by all algorithms, the OptimizeOptions
 * for configuring convergence criteria, and the convergence checking logic.
 *
 * @node result-types
 * @contract result-types.test.ts
 * @hint types: Use plain objects/structs, not classes. Keep it simple.
 * @hint defaults: Our defaults differ from scipy (1e-5) and MATLAB (1e-6).
 *       We chose 1e-8 for gradient tolerance, matching Optim.jl.
 *       Document provenance of each default.
 */

/**
 * Configuration options for optimization.
 *
 * Default gradient tolerance: 1e-8 (matches Optim.jl; scipy uses 1e-5, MATLAB uses 1e-6).
 * Default step tolerance: 1e-8 (matches MATLAB; Optim.jl disables by default).
 * Default function tolerance: 1e-12 (stricter than most; catches near-convergence stalls).
 * Default max iterations: 1000 (matches Optim.jl; MATLAB uses 400).
 */
export interface OptimizeOptions {
  /** Gradient infinity-norm tolerance. Converged when ||g||_inf < gradTol. */
  gradTol: number;
  /** Step size tolerance. Converged when ||x_new - x_old||_inf < stepTol. */
  stepTol: number;
  /** Function value change tolerance. Converged when |f_new - f_old| < funcTol. */
  funcTol: number;
  /** Maximum number of iterations. */
  maxIterations: number;
}

/** Create default options, optionally overriding specific fields. */
export function defaultOptions(overrides?: Partial<OptimizeOptions>): OptimizeOptions {
  return {
    gradTol: 1e-8,
    stepTol: 1e-8,
    funcTol: 1e-12,
    maxIterations: 1000,
    ...overrides,
  };
}

/**
 * Result of an optimization run.
 *
 * Modeled after scipy's OptimizeResult and Optim.jl's MultivariateOptimizationResults.
 * Uses plain fields rather than accessor methods for simplicity.
 */
export interface OptimizeResult {
  /** Solution vector (minimizer). */
  x: number[];
  /** Objective function value at solution. */
  fun: number;
  /** Gradient at solution (if available). */
  gradient: number[] | null;
  /** Number of iterations performed. */
  iterations: number;
  /** Number of objective function evaluations. */
  functionCalls: number;
  /** Number of gradient evaluations. */
  gradientCalls: number;
  /** Whether the optimizer converged (met a convergence criterion). */
  converged: boolean;
  /** Human-readable description of termination reason. */
  message: string;
}

/**
 * Why did the optimizer stop?
 *
 * Used internally to determine `converged` and `message` in OptimizeResult.
 */
export type ConvergenceReason =
  | { kind: "gradient"; gradNorm: number }
  | { kind: "step"; stepNorm: number }
  | { kind: "function"; funcChange: number }
  | { kind: "maxIterations"; iterations: number }
  | { kind: "lineSearchFailed"; message: string };

/**
 * Check convergence criteria. Returns the first matching reason, or null
 * if no criterion is met.
 *
 * Checks in order: gradient → step → function value → max iterations.
 * This order matches Optim.jl's convention (gradient is primary).
 */
export function checkConvergence(
  gradNorm: number,
  stepNorm: number,
  funcChange: number,
  iteration: number,
  options: OptimizeOptions,
): ConvergenceReason | null {
  if (gradNorm < options.gradTol) {
    return { kind: "gradient", gradNorm };
  }
  if (stepNorm < options.stepTol) {
    return { kind: "step", stepNorm };
  }
  if (funcChange < options.funcTol) {
    return { kind: "function", funcChange };
  }
  if (iteration >= options.maxIterations) {
    return { kind: "maxIterations", iterations: iteration };
  }
  return null;
}

/** Is this convergence reason considered "converged" (vs "stopped")? */
export function isConverged(reason: ConvergenceReason): boolean {
  return (
    reason.kind === "gradient" ||
    reason.kind === "step" ||
    reason.kind === "function"
  );
}

/** Human-readable message for a convergence reason. */
export function convergenceMessage(reason: ConvergenceReason): string {
  switch (reason.kind) {
    case "gradient":
      return `Converged: gradient norm ${reason.gradNorm.toExponential(2)} below tolerance`;
    case "step":
      return `Converged: step size ${reason.stepNorm.toExponential(2)} below tolerance`;
    case "function":
      return `Converged: function change ${reason.funcChange.toExponential(2)} below tolerance`;
    case "maxIterations":
      return `Stopped: reached maximum iterations (${reason.iterations})`;
    case "lineSearchFailed":
      return `Stopped: line search failed (${reason.message})`;
  }
}

File: research/experiments/sweep/runs/opt-medium-csharp-opus45/src/ResultTypes.cs
using System;

namespace Optimization
{
    public class OptimizeOptions
    {
        public double GradTol { get; set; } = 1e-8;
        public double StepTol { get; set; } = 1e-8;
        public double FuncTol { get; set; } = 1e-12;
        public int MaxIterations { get; set; } = 1000;
    }

    public class OptimizeResult
    {
        public double[] X { get; set; } = Array.Empty<double>();
        public double Fun { get; set; }
        public double[]? Gradient { get; set; }
        public int Iterations { get; set; }
        public int FunctionCalls { get; set; }
        public int GradientCalls { get; set; }
        public bool Converged { get; set; }
        public string Message { get; set; } = "";
    }

    public enum ConvergenceKind
    {
        Gradient,
        Step,
        Function,
        MaxIterations,
        LineSearchFailed
    }

    public class ConvergenceReason
    {
        public ConvergenceKind Kind { get; }

        public ConvergenceReason(ConvergenceKind kind)
        {
            Kind = kind;
        }
    }

    public static class ResultTypes
    {
        public static OptimizeOptions DefaultOptions(OptimizeOptions? overrides = null)
        {
            var opts = new OptimizeOptions();
            if (overrides != null)
            {
                // Apply overrides by checking if they differ from defaults
                // Since we can't distinguish "not set" from "set to default" with value types,
                // we just copy all fields from overrides
                opts.GradTol = overrides.GradTol;
                opts.StepTol = overrides.StepTol;
                opts.FuncTol = overrides.FuncTol;
                opts.MaxIterations = overrides.MaxIterations;
            }
            return opts;
        }

        /// <summary>
        /// Check convergence in order: gradient, step, function, maxIterations.
        /// Returns null if no criterion is met.
        /// </summary>
        public static ConvergenceReason? CheckConvergence(
            double gradNorm, double stepNorm, double funcChange,
            int iteration, OptimizeOptions opts)
        {
            if (gradNorm < opts.GradTol)
                return new ConvergenceReason(ConvergenceKind.Gradient);
            if (stepNorm < opts.StepTol)
                return new ConvergenceReason(ConvergenceKind.Step);
            if (funcChange < opts.FuncTol)
                return new ConvergenceReason(ConvergenceKind.Function);
            if (iteration >= opts.MaxIterations)
                return new ConvergenceReason(ConvergenceKind.MaxIterations);
            return null;
        }

        public static bool IsConverged(ConvergenceReason reason)
        {
            return reason.Kind == ConvergenceKind.Gradient
                || reason.Kind == ConvergenceKind.Step
                || reason.Kind == ConvergenceKind.Function;
        }

        public static string ConvergenceMessage(ConvergenceReason reason)
        {
            return reason.Kind switch
            {
                ConvergenceKind.Gradient => "Converged: gradient norm below tolerance",
                ConvergenceKind.Step => "Converged: step size below tolerance",
                ConvergenceKind.Function => "Converged: function change below tolerance",
                ConvergenceKind.MaxIterations => "Stopped: reached maximum iterations",
                ConvergenceKind.LineSearchFailed => "Stopped: line search failed",
                _ => "Unknown convergence reason"
            };
        }
    }
}
